### DISCUSSION TITLE
Is there a way to do Single Table Inheritance?

### DISCUSSION THREAD

**Discussion by grrenier** [noam-honig/radweb#193] [Q&A]
Hello @noam-honig, I'm wondering if there a way to do something similar to what is done in Doctrine ORM for the Single Table Inheritance pattern?

```
<?php
namespace MyProject\Model;

#[Entity]
#[InheritanceType('SINGLE_TABLE')]
#[DiscriminatorColumn(name: 'discr', type: 'string')]
#[DiscriminatorMap(['person' => Person::class, 'employee' => Employee::class])]
class Person
{
    // ...
}

#[Entity]
class Employee extends Person
{
    // ...
}

```

**Comment by noam-honig** [comment]
Try out the following code, I think it'll achieve what you were looking for:
```ts
import { Entity, Fields } from 'remult'

@Entity('people', {
  allowApiCrud: true,
})
export class Person {
  @Fields.cuid()
  id = ''
  @Fields.string()
  name = ''
  @Fields.string({ allowApiUpdate:false })
  discr = ''
}

@Entity<Person>('employees', {
  allowApiCrud: true,
  dbName: 'people',
  backendPrefilter: {
    discr: 'person',
  },
  saving: (self) => {
    self.discr = 'person'
  },
})
export class Employee extends Person {}

```

On the `Employee` entity I've used the `dbName` option to set the same table name.

I've added a `backendPrefilter` that will make sure that any query to that entity will add the filter of `discr = 'person'`, 
And added a `saving` handler that'll update the field `discr` with the correct value.

Let me know what you think

**Comment by grrenier** [comment]
Yes, it should work, maybe you can also add a `backendPrefilter` to the parent to get the children entities.

**Reply by yoni-rapoport** [reply]
Actually, if you query the `Person` entity directly you're simply going to get the data for all the inheriting entities. No need for a `backendPrefilter` there.

**Reply by grrenier** [reply]
You're right, in that case it's not required, but for example when you have 2 types of employees `EmployeeA` and `EmployeeB` that inherit from `Employee` it will be required at the employee level to be able to get all the employees when you query the `Employee` entity.

**Reply by noam-honig** [reply]
I think you can achieve that with an 'or' or 'in' statement. (I see now that in my original response, I confused 'person' with 'employee' in the `discr` field the code :)

Let me know if this works for you:
```ts
import { Entity, Fields } from 'remult'

@Entity('people', {
  allowApiCrud: true,
})
export class Person {
  @Fields.cuid()
  id = ''
  @Fields.string()
  name = ''
  @Fields.string({ allowApiUpdate:false })
  discr = ''
}

@Entity<Person>('employees', {
  allowApiCrud: true,
  dbName: 'people',
  backendPrefilter: {
    discr: ["employeeA","employeeB"],
  },
})
export class Employee extends Person {}

@Entity<Person>('employeesA', {
  allowApiCrud: true,
  dbName: 'people',
  backendPrefilter: {
    discr: "employeeA",
  },
  saving: (self) => {
    self.discr = 'employeeA'
  },
})
export class EmployeeA extends Employee {}

@Entity<Person>('employeesB', {
  allowApiCrud: true,
  dbName: 'people',
  backendPrefilter: {
    discr: "employeeB",
  },
  saving: (self) => {
    self.discr = 'employeeB'
  },
})
export class EmployeeB extends Employee {}
```

**Comment by elier** [comment]
It would be great to retrieve the instances of the corresponding discriminator value rather than the type used in the query.

The following code retrieves Person instances rather than the instance indicated by the discr. That is forcing me to query the database again and again using the right type every time. 
```
const people = await remult.repo(Person).find();
assert(type of people[0] == 'EmployeeA') // => fails
assert(type of people[1] == 'EmployeeB') // => fails
```

Another problem that I'm facing here is that Person cannot be abstract. 

Also, when using constructors the value of discr is wrongly set. It looks like the discr assignation is based on attribute matching rather than the type provided in the insert.

Ex: 
```
remult.repo(Customer).insert({firstName: '', lastName: ''});
```

When having a constructor in Customer and Person, discr take the value of "person". I think because it is Person who has the attributes firstName and lastName, but the provided type was Customer, I expect the discr take "customer" value. Or the constructor is messing up something.

**Comment by elier** [comment]
@noam-honig a couple of potential bugs here

**Comment by noam-honig** [comment]
@elier, 

When you use `remult.repo(Person).find();`  it'll return an instance of Person per the current design.

You can create follow up code that'll load the specific instance based on the `discr` column - but that is not built in or planned at this stage.

As for the second issue you've mentioned with customer - I didn't exactly understand the issue - please add a code sample
